Accessibility checks for methods with an InvokeDynamic instruction

lrytz opened the pull request:
Implements the necessary tests to check if a method with an
InvokeDynamic instruction can be inlined into a destination class.

Only InvokeDynamic instructions with LambdaMetaFactory as bootstrap
methods can be inlined. The accessibility checks cannot be implemented
generically, because it depends on what the bootstrap method is doing.
In particular, the bootstrap method receives a Lookup object as
argument which can be used to access private methods of the class
where the InvokeDynamic method is located.

A comment in the inliner explains the details.

lrytz committed:
Set -Xsource to 2.12 by default

lrytz committed:
Warn when combining settings for the old optimizer with -YGenBCode

Remove some unnecessary flags files
 - neg/t4425.flags
 - run/blame_eye_triple_eee-double.flags
 - run/blame_eye_triple_eee-float.flags

Force tests that use -optimize to GenASM
 - neg/sealed-final-neg.flags
 - pos/inline-access-levels.flags
 - pos/inliner2.flags
 - pos/sealed-final.flags
 - pos/t3420.flags
 - pos/t8410.flags
 - run/constant-optimization.flags
 - run/dead-code-elimination.flags
 - run/elidable-opt.flags
 - run/finalvar.flags
 - run/icode-reader-dead-code.scala
 - run/optimizer-array-load.flags
 - run/synchronized.flags
 - run/t3509.flags
 - run/t3569.flags
 - run/t4285.flags
 - run/t4935.flags
 - run/t5789.scala
 - run/t6188.flags
 - run/t7459b-optimize.flags
 - run/t7582.flags
 - run/t7582b.flags
 - run/t8601.flags
 - run/t8601b.flags
 - run/t8601c.flags
 - run/t8601d.flags
 - run/t8601e.flags
 - run/t9003.flags

Move some tests to the new optimizer
 - run/classfile-format-51.scala
 - run/classfile-format-52.scala
 - run/run-bug4840.flags
 - run/t2106.flags
 - run/t6102.flags

lrytz committed:
Enable the new optimizer when building scala

No change in build.sbt, there's no optimizer settings there yet.

Ignore the inliner warning in presentation/t7678 and run/t8029.scala,
noted in https://issues.scala-lang.org/browse/SI-9378

lrytz committed:
SI-9376 don't crash when inlining a closure body that throws.

If the closure body method has return type Nothing$, add an `ATHROW`
instruction after the callsite. This is required for computing stack
map frames, as explained in a comment in BCodeBodyBuilder.adapt.

Similar for closure bodies with return type Null$.

lrytz committed:
Prevent infinite recursion in ProdConsAnalyzer

When an instruction is its own producer or consumer, the
`initialProducer` / `ultimateConsumer` methods would loop.
While loops or @tailrec annotated methods can generate such bytecode.

lrytz committed:
Skip mirror class when invoking deserializeLambda

Generate the invocation to LambdaDeserializer.deserializeLambda by
loading the static MODULE$ field instead of calling the static method
in the mirror class.

This is more scala-y. Also, mirror classes don't have an InlineInfo
classfile attribute, so the inliner would yield a warning about
the mirror class callsite.

Also skip the stack map frame instruction - frames are computed by
the ams classfile writer.

lrytz committed:
Accessibility checks for methods with an InvokeDynamic instruction

Implements the necessary tests to check if a method with an
InvokeDynamic instruction can be inlined into a destination class.

Only InvokeDynamic instructions with LambdaMetaFactory as bootstrap
methods can be inlined. The accessibility checks cannot be implemented
generically, because it depends on what the bootstrap method is doing.
In particular, the bootstrap method receives a Lookup object as
argument which can be used to access private methods of the class
where the InvokeDynamic method is located.

A comment in the inliner explains the details.

adriaanm commented on src/compiler/scala/tools/nsc/backend/jvm/opt/Inliner.scala:
```diff
@@ -687,9 +688,67 @@ class Inliner[BT <: BTypes](val btypes: BT) {
+
+        if (destinationClass == calleeDeclarationClass) {
+          Right(true) // within the same class, any indy instruction can be inlined
+        } else if (closureOptimizer.isClosureInstantiation(indy)) {
```
Could we avoid the casts of `bsmArgs`, here and in `ClosureOptimizer`, by refactoring `isClosureInstantiation` to something like `def isLambdaMetaFactoryCall(indy: InvokeDynamicInsnNode): LambdaMetaFactoryCall` with `case class LambdaMetaFactoryCall(samMethodType: Type, implMethod: Handle, instantiatedMethodType: Type)`? Or perhaps `isLambdaMetaFactoryCall` should be the `unapply` of `LambdaMetaFactoryCall`.

adriaanm replied:
If you end up going with the extractor, this `if` could be lifted to the outer `match`.

adriaanm commented on src/partest-extras/scala/tools/partest/ASMConverters.scala:
```diff
@@ -119,6 +123,13 @@ object ASMConverters {
       case i: t.LineNumberNode         => LineNumber   (i.line, applyLabel(i.start))
     }
 
+    private def convertBsmArgs(a: Array[Object]): List[Object] = a.map({
```
Unnecessary parens?

lrytz replied:
I left this as is because the second parameter list `(collection.breakOut)` looks strange otherwise.

adriaanm commented on src/partest-extras/scala/tools/partest/ASMConverters.scala:
```diff
@@ -197,21 +208,28 @@ object ASMConverters {
   }
 
+  def unconvertMethodHandle(h: MethodHandle): asm.Handle = new asm.Handle(h.tag, h.owner, h.name, h.desc)
+  def unconvertBsmArgs(a: List[Object]): Array[Object] = a.map({
```
Unnecessary parens?

lrytz replied:
same here.

adriaanm commented on test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala:
```diff
@@ -991,4 +991,26 @@ class InlinerTest extends ClearAfterClass {
+        |    val f = (x: String) => x.trim
+        |    f(s)
+        |  }
+        |  def t = m("foo")
```
So, this is the `destinationClass == calleeDeclarationClass` case, right? Could you also add a test for the inter-class case?

lrytz committed:
Fix bytecode stability

When there are multiple closure allocations and invocations in the
same method, ensure that the callsites are re-written to the body
methods in a consistent order. Otherwsie the bytecode is not stable
(the local variable indices depend on the order in which the calls
are re-written)

lrytz committed:
Small refactoring to the closure optimizer

Introduces an extractor `LMFInvokeDynamic` that matches InvokeDynamic
instructions that are LambdaMetaFactory calls.

The case class `LambdaMetaFactoryCall` holds such an InvokeDynamic
instruction. It also holds the bootstrap arguments (samMethodType,
implMethod, instantiatedMethodType) so that they can be accessed
without casting the indy.bsmArgs.

The `closureInstantiations` map in the call graph now stores
ClosureInstantiation objects instead of a tuple.

This simplifies some code and gets rid of a few casts.

lrytz commented:
Fixed a number of things:
- rebased on top of #4602, so we can [revert some partest filters](https://github.com/scala/scala/commit/e9c742e1d27c41ee8646ea20dfb59efbc3d94ef3#diff-f38403c4df9819d1fd49062815b9d46f) that were introduced there
- made sure that closure invocations are re-written to the body methods in a consistent order, to ensure bytecode stability ([first commit](https://github.com/scala/scala/commit/4e9be26c1d0ea4818b54c1f882a2972f439b6e39))
- re-factored `isClosureInstantiation` into an extractor, introduced some classes to clean up the code ([third commit](https://github.com/scala/scala/commit/6e273f7cb6a0f6ff12d899676d30aff5d83b3bb1))
- added a test for inlining of `invokedynamic` instructions across classes
- left the second commit mostly as-is, to simplify reviewing

lrytz commented:
/rebuild

lrytz commented:
/rebuild

adriaanm commented:
/synch

adriaanm committed:
Integrate the LMFInvokeDynamic extractor into LambdaMetaFactoryCall

adriaanm commented:
I implemented my suggestion at https://github.com/lrytz/scala/compare/inlineIndy...adriaanm:lrytz-inlineIndy?expand=1 (CI: https://scala-ci.typesafe.com/job/scala-2.12.x-validate-main/220/console)

retronym commented:
Could do with a couple of extra tests that go through the no-inline paths, after that LGTM.

lrytz committed:
Test warnings when an indy cannot be inlined

There are two cases tested here
  - An indyLMF where the lambda body method is private
  - An indy where the bootstrap method is not LMF

lrytz commented:
@adriaanm pushed your changes
@retronym pushed tests for the no-inline paths

adriaanm commented:
LGTM! (Although somehow my favorite Scala symbol got expunged: `_@_*`.)

retronym commented:
> Although somehow my favorite Scala symbol got expunged: `_@_*`

🐌?

adriaanm commented:
Exactamundo.

lrytz commented:
scalac should accept `\u1F40C` as `_@_*`.
